File: macOS/GTABridge/Networking/BridgeClientSession.swift
@preconcurrency import Network
import Foundation
import GTAControlCore
import OSLog

@MainActor
protocol BridgeClientSessionDelegate: AnyObject {
    func session(_ session: BridgeClientSession, receivedHello hello: HelloMessage)
    func session(_ session: BridgeClientSession, authenticatedClient clientID: UUID)
    func session(_ session: BridgeClientSession, requestedPairing request: PairingRequest)
    func session(_ session: BridgeClientSession, receivedCommand command: CommandMessage) async
    func session(_ session: BridgeClientSession, receivedGodModeCommand command: GodModeCommandMessage) async
    func session(_ session: BridgeClientSession, receivedWreckPreservationCommand command: WreckPreservationCommandMessage) async
    func session(_ session: BridgeClientSession, failedWith error: Error)
    func sessionDidDisconnect(_ session: BridgeClientSession)
}

enum BridgeSessionError: Error, Equatable, LocalizedError {
    case controllerHelloRequired
    case helloRequired
    case pairingRequired
    case clientIdentityMismatch
    case unexpectedCleartextMessage
    case unexpectedSecureMessage
    case heartbeatRequired
    case malformedEncryptedPacket
    case connectionClosed

    var errorDescription: String? {
        switch self {
        case .controllerHelloRequired:
            "Il peer non si e identificato come controller."
        case .helloRequired:
            "Il client deve inviare hello prima degli altri messaggi."
        case .pairingRequired:
            "Il client non e associato a questo Mac."
        case .clientIdentityMismatch:
            "L'identita del client non coincide con la sessione."
        case .unexpectedCleartextMessage:
            "Messaggio non consentito in chiaro."
        case .unexpectedSecureMessage:
            "Messaggio cifrato non consentito in questa fase."
        case .heartbeatRequired:
            "Il client deve completare la prova cifrata prima di inviare comandi."
        case .malformedEncryptedPacket:
            "Pacchetto cifrato non valido."
        case .connectionClosed:
            "Connessione chiusa dal client."
        }
    }
}

@MainActor
final class BridgeClientSession: Identifiable {
    let id = UUID()
    private(set) var clientID: UUID?
    private var presentedClientID: UUID?

    private static let logger = Logger(
        subsystem: "com.matteozampieri.GTABridge",
        category: "ClientSession"
    )

    private let connection: NWConnection
    private let bridgeID: UUID
    private let sessionStore: KeychainSessionStore
    private weak var delegate: (any BridgeClientSessionDelegate)?
    private var frameDecoder = FrameDecoder()
    private var secureChannel: SecureChannel?
    private var hasAuthenticatedSecurePacket = false
    private var nextOutgoingSecureSequence: UInt64 = 1
    private var isCancelled = false

    init(
        connection: NWConnection,
        bridgeID: UUID,
        sessionStore: KeychainSessionStore,
        delegate: any BridgeClientSessionDelegate
    ) {
        self.connection = connection
        self.bridgeID = bridgeID
        self.sessionStore = sessionStore
        self.delegate = delegate
    }

    func start() {
        connection.stateUpdateHandler = { [weak self] state in
            Task { @MainActor in
                self?.handleConnectionState(state)
            }
        }
        connection.start(queue: .main)
        receiveNextChunk()
    }

    func cancel() {
        guard !isCancelled else { return }
        isCancelled = true
        connection.cancel()
    }

    func activateSecureChannel(sessionKey: SessionKey, clientID: UUID) {
        self.clientID = clientID
        secureChannel = SecureChannel(sessionKey: sessionKey)
    }

    func sendPairingChallenge(_ challenge: PairingChallenge) async throws {
        try await sendClear(.pairingChallenge(challenge))
    }

    func sendPairingApproval(_ approval: PairingApproval) async throws {
        try await sendClear(.pairingApproval(approval))
    }

    func sendAcknowledgement(_ acknowledgement: AcknowledgementMessage) async throws {
        try await sendSecure(.acknowledgement(acknowledgement))
    }

    func sendTrainerState(_ snapshot: TrainerStateSnapshot) async throws {
        guard hasAuthenticatedSecurePacket else { return }
        try await sendSecure(.trainerState(TrainerStateMessage(snapshot: snapshot)))
    }

    func sendError(_ error: ProtocolErrorMessage, securely: Bool) async {
        do {
            if securely {
                try await sendSecure(.error(error))
            } else {
                try await sendClear(.error(error))
            }
        } catch {
            delegate?.session(self, failedWith: error)
        }
    }

    private func receiveNextChunk() {
        guard !isCancelled else { return }
        connection.receive(minimumIncompleteLength: 1, maximumLength: ProtocolConstants.maximumFrameBytes + 4) {
            [weak self] data, _, isComplete, error in
            Task { @MainActor in
                guard let self, !self.isCancelled else { return }

                if let data, !data.isEmpty {
                    do {
                        let frames = try self.frameDecoder.append(data)
                        for frame in frames {
                            try await self.handleFrame(frame)
                        }
                    } catch {
                        self.delegate?.session(self, failedWith: error)
                        self.cancel()
                        return
                    }
                }

                if let error {
                    self.delegate?.session(self, failedWith: error)
                    self.cancel()
                    return
                }

                if isComplete {
                    self.delegate?.session(self, failedWith: BridgeSessionError.connectionClosed)
                    self.cancel()
                    return
                }

                self.receiveNextChunk()
            }
        }
    }

    private func handleFrame(_ payload: Data) async throws {
        if clientID == nil {
            let message = try ProtocolCodec.decode(payload)
            try await handleClearMessage(message)
            return
        }

        // The controller Keychain can be reset independently from the Mac Keychain,
        // for example after reinstalling the iPhone app. Before the first authenticated
        // packet, allow that same announced client to restart pairing explicitly.
        if !hasAuthenticatedSecurePacket,
           let clearMessage = try? ProtocolCodec.decode(payload),
           case let .pairingRequest(request) = clearMessage,
           request.clientID == presentedClientID,
           let staleClientID = clientID {
            try sessionStore.removeSessionKey(for: staleClientID)
            clientID = nil
            secureChannel = nil
            nextOutgoingSecureSequence = 1
            try await handleClearMessage(clearMessage)
            return
        }

        guard var channel = secureChannel else {
            throw BridgeSessionError.pairingRequired
        }

        let packet: EncryptedPacket
        do {
            packet = try JSONDecoder().decode(EncryptedPacket.self, from: payload)
        } catch {
            throw BridgeSessionError.malformedEncryptedPacket
        }

        do {
            let message = try channel.open(packet)
            secureChannel = channel
            try await handleSecureMessage(message)
        } catch {
            secureChannel = channel
            throw error
        }
    }

    private func handleClearMessage(_ message: WireMessage) async throws {
        switch message {
        case let .hello(hello):
            guard hello.role == .controller else {
                throw BridgeSessionError.controllerHelloRequired
            }
            guard presentedClientID == nil || presentedClientID == hello.peerID else {
                throw BridgeSessionError.clientIdentityMismatch
            }

            presentedClientID = hello.peerID
            try await sendClear(.hello(HelloMessage(peerID: bridgeID, role: .bridge)))

            if let storedKey = try sessionStore.sessionKey(for: hello.peerID) {
                activateSecureChannel(sessionKey: storedKey, clientID: hello.peerID)
            }
            delegate?.session(self, receivedHello: hello)

        case let .pairingRequest(request):
            guard clientID == nil, request.clientID == presentedClientID else {
                throw BridgeSessionError.unexpectedCleartextMessage
            }
            delegate?.session(self, requestedPairing: request)

        default:
            throw clientID == nil
                ? BridgeSessionError.helloRequired
                : BridgeSessionError.unexpectedCleartextMessage
        }
    }

    private func handleSecureMessage(_ message: WireMessage) async throws {
        switch message {
        case let .command(command):
            guard hasAuthenticatedSecurePacket else {
                throw BridgeSessionError.heartbeatRequired
            }
            guard command.envelope.clientID == clientID else {
                throw BridgeSessionError.clientIdentityMismatch
            }
            await delegate?.session(self, receivedCommand: command)
        case let .godModeCommand(command):
            guard hasAuthenticatedSecurePacket else {
                throw BridgeSessionError.heartbeatRequired
            }
            guard command.clientID == clientID else {
                throw BridgeSessionError.clientIdentityMismatch
            }
            await delegate?.session(self, receivedGodModeCommand: command)
        case let .wreckPreservationCommand(command):
            guard hasAuthenticatedSecurePacket else {
                throw BridgeSessionError.heartbeatRequired
            }
            guard command.clientID == clientID else {
                throw BridgeSessionError.clientIdentityMismatch
            }
            await delegate?.session(self, receivedWreckPreservationCommand: command)
        case .heartbeat:
            try await sendHeartbeatResponse()
            if !hasAuthenticatedSecurePacket, let clientID {
                hasAuthenticatedSecurePacket = true
                delegate?.session(self, authenticatedClient: clientID)
            }
        default:
            throw BridgeSessionError.unexpectedSecureMessage
        }
    }

    private func sendClear(_ message: WireMessage) async throws {
        try await send(ProtocolCodec.encodeFrame(message))
    }

    private func sendSecure(_ message: WireMessage) async throws {
        guard var channel = secureChannel else {
            throw BridgeSessionError.pairingRequired
        }
        let packet = try channel.seal(message)
        secureChannel = channel
        if nextOutgoingSecureSequence == UInt64.max {
            nextOutgoingSecureSequence = 0
        } else {
            nextOutgoingSecureSequence += 1
        }
        try await send(FrameCodec.encode(packet))
    }

    private func sendHeartbeatResponse() async throws {
        guard nextOutgoingSecureSequence > 0 else {
            throw SecureChannelError.exhaustedSequence
        }
        try await sendSecure(
            .heartbeat(HeartbeatMessage(sequence: nextOutgoingSecureSequence))
        )
    }

    private func send(_ data: Data) async throws {
        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, any Error>) in
            connection.send(content: data, completion: .contentProcessed { error in
                if let error {
                    continuation.resume(throwing: error)
                } else {
                    continuation.resume(returning: ())
                }
            })
        }
    }

    private func handleConnectionState(_ state: NWConnection.State) {
        switch state {
        case .failed(let error):
            Self.logger.error("Client connection failed: \(error.localizedDescription, privacy: .public)")
            delegate?.session(self, failedWith: error)
            cancel()
        case .waiting(let error):
            Self.logger.info("Client connection waiting: \(error.localizedDescription, privacy: .public)")
        case .cancelled:
            delegate?.sessionDidDisconnect(self)
        default:
            break
        }
    }
}

File: Windows/GTABridge.Windows/Networking/BridgeClientSession.cs
using System.Net.Sockets;
using System.Text.Json;
using GTAControlCore.Windows;
using GTABridge.Windows.Storage;

namespace GTABridge.Windows.Networking;

internal sealed class BridgeClientSession : IAsyncDisposable
{
    private readonly TcpClient _client;
    private readonly NetworkStream _stream;
    private readonly Guid _bridgeID;
    private readonly ProtectedSessionStore _sessionStore;
    private readonly BridgeServer _server;
    private readonly FrameDecoder _decoder = new();
    private readonly SemaphoreSlim _sendGate = new(1, 1);
    private readonly object _channelGate = new();
    private SecureChannel? _secureChannel;
    private Guid? _presentedClientID;
    private bool _authenticated;
    private int _disposed;

    public BridgeClientSession(
        TcpClient client,
        Guid bridgeID,
        ProtectedSessionStore sessionStore,
        BridgeServer server)
    {
        _client = client;
        _client.NoDelay = true;
        _stream = client.GetStream();
        _bridgeID = bridgeID;
        _sessionStore = sessionStore;
        _server = server;
    }

    public Guid ID { get; } = Guid.NewGuid();
    public Guid? ClientID { get; private set; }
    public bool IsAuthenticated => _authenticated;

    public async Task RunAsync(CancellationToken cancellationToken)
    {
        var buffer = new byte[ProtocolConstants.MaximumFrameBytes + 4];
        while (!cancellationToken.IsCancellationRequested)
        {
            var count = await _stream.ReadAsync(buffer, cancellationToken).ConfigureAwait(false);
            if (count == 0) return;
            foreach (var frame in _decoder.Append(buffer.AsSpan(0, count)))
                await HandleFrameAsync(frame, cancellationToken).ConfigureAwait(false);
        }
    }

    public void ActivateSecureChannel(byte[] sessionKey, Guid clientID)
    {
        ClientID = clientID;
        lock (_channelGate)
        {
            _secureChannel?.Dispose();
            _secureChannel = new SecureChannel(sessionKey);
        }
    }

    public Task SendPairingChallengeAsync(PairingChallenge challenge, CancellationToken cancellationToken) =>
        SendClearAsync(WireMessage.PairingChallenge(challenge), cancellationToken);

    public Task SendPairingApprovalAsync(PairingApproval approval, CancellationToken cancellationToken) =>
        SendClearAsync(WireMessage.PairingApproval(approval), cancellationToken);

    public Task SendAcknowledgementAsync(AcknowledgementMessage acknowledgement, CancellationToken cancellationToken) =>
        SendSecureAsync(WireMessage.Acknowledgement(acknowledgement), cancellationToken);

    public async Task SendTrainerStateAsync(TrainerStateSnapshot snapshot, CancellationToken cancellationToken)
    {
        if (_authenticated)
            await SendSecureAsync(WireMessage.TrainerState(snapshot), cancellationToken).ConfigureAwait(false);
    }

    private async Task HandleFrameAsync(byte[] frame, CancellationToken cancellationToken)
    {
        if (ClientID is null)
        {
            await HandleClearAsync(ProtocolJson.Decode(frame), cancellationToken).ConfigureAwait(false);
            return;
        }

        if (!_authenticated && TryDecodePairingRequest(frame, out var request) && request.ClientID == _presentedClientID)
        {
            _sessionStore.Remove(ClientID.Value);
            ClientID = null;
            lock (_channelGate)
            {
                _secureChannel?.Dispose();
                _secureChannel = null;
            }
            await HandleClearAsync(new WireMessage("pairingRequest", request), cancellationToken).ConfigureAwait(false);
            return;
        }

        WireMessage message;
        lock (_channelGate)
        {
            if (_secureChannel is null) throw new InvalidDataException(LocalizedText.Choose(
                "Pairing richiesto.",
                "Pairing required."));
            message = _secureChannel.Open(frame);
        }
        await HandleSecureAsync(message, cancellationToken).ConfigureAwait(false);
    }

    private async Task HandleClearAsync(WireMessage message, CancellationToken cancellationToken)
    {
        switch (message.Payload)
        {
            case HelloMessage hello:
                if (hello.Role != PeerRole.Controller) throw new InvalidDataException(LocalizedText.Choose(
                    "Il peer non è un controller.",
                    "The peer is not a controller."));
                if (_presentedClientID is not null && _presentedClientID != hello.PeerID)
                    throw new InvalidDataException(LocalizedText.Choose(
                        "L'identità del client è cambiata durante la connessione.",
                        "The client identity changed during the connection."));
                _presentedClientID = hello.PeerID;
                await SendClearAsync(WireMessage.Hello(_bridgeID, PeerRole.Bridge), cancellationToken).ConfigureAwait(false);
                var storedKey = _sessionStore.Load(hello.PeerID);
                if (storedKey is not null)
                {
                    try
                    {
                        ActivateSecureChannel(storedKey, hello.PeerID);
                    }
                    finally
                    {
                        System.Security.Cryptography.CryptographicOperations.ZeroMemory(storedKey);
                    }
                    _server.OnAuthenticating(hello.PeerID);
                }
                else
                {
                    _server.OnPairingNeeded();
                }
                break;

            case PairingRequest pairingRequest when ClientID is null && pairingRequest.ClientID == _presentedClientID:
                await _server.RequestPairingAsync(this, pairingRequest, cancellationToken).ConfigureAwait(false);
                break;

            default:
                throw new InvalidDataException(_presentedClientID is null
                    ? LocalizedText.Choose("Il client deve inviare hello per primo.", "The client must send hello first.")
                    : LocalizedText.Choose("Messaggio in chiaro non consentito.", "Clear-text message not allowed."));
        }
    }

    private async Task HandleSecureAsync(WireMessage message, CancellationToken cancellationToken)
    {
        switch (message.Payload)
        {
            case HeartbeatMessage:
                await SendHeartbeatAsync(cancellationToken).ConfigureAwait(false);
                if (!_authenticated && ClientID is { } clientID)
                {
                    _authenticated = true;
                    await _server.OnAuthenticatedAsync(this, clientID, cancellationToken).ConfigureAwait(false);
                }
                break;

            case CommandMessage command:
                RequireAuthenticated(command.Envelope.ClientID);
                await _server.HandleCommandAsync(this, command, cancellationToken).ConfigureAwait(false);
                break;

            case GodModeCommandMessage command:
                RequireAuthenticated(command.ClientID);
                await _server.HandleGodModeAsync(this, command, cancellationToken).ConfigureAwait(false);
                break;

            case WreckPreservationCommandMessage command:
                RequireAuthenticated(command.ClientID);
                await _server.HandleWreckPreservationAsync(this, command, cancellationToken).ConfigureAwait(false);
                break;

            default:
                throw new InvalidDataException(LocalizedText.Choose(
                    "Messaggio cifrato non consentito.",
                    "Encrypted message not allowed."));
        }
    }

    private void RequireAuthenticated(Guid clientID)
    {
        if (!_authenticated) throw new InvalidDataException(LocalizedText.Choose(
            "Heartbeat cifrato richiesto prima dei comandi.",
            "An encrypted heartbeat is required before commands."));
        if (clientID != ClientID) throw new InvalidDataException(LocalizedText.Choose(
            "Identità client non valida.",
            "Invalid client identity."));
    }

    private Task SendHeartbeatAsync(CancellationToken cancellationToken)
    {
        EncryptedPacket packet;
        lock (_channelGate)
        {
            if (_secureChannel is null) throw new InvalidDataException(LocalizedText.Choose(
                "Canale cifrato assente.",
                "Encrypted channel is unavailable."));
            packet = _secureChannel.Seal(WireMessage.Heartbeat(_secureChannel.NextSendingSequence));
        }
        var payload = JsonSerializer.SerializeToUtf8Bytes(packet, ProtocolJson.Options);
        return SendFrameAsync(FrameCodec.Encode(payload), cancellationToken);
    }

    private Task SendClearAsync(WireMessage message, CancellationToken cancellationToken) =>
        SendFrameAsync(FrameCodec.Encode(ProtocolJson.Encode(message)), cancellationToken);

    private Task SendSecureAsync(WireMessage message, CancellationToken cancellationToken)
    {
        EncryptedPacket packet;
        lock (_channelGate)
        {
            if (_secureChannel is null) throw new InvalidDataException(LocalizedText.Choose(
                "Canale cifrato assente.",
                "Encrypted channel is unavailable."));
            packet = _secureChannel.Seal(message);
        }
        var payload = JsonSerializer.SerializeToUtf8Bytes(packet, ProtocolJson.Options);
        return SendFrameAsync(FrameCodec.Encode(payload), cancellationToken);
    }

    private async Task SendFrameAsync(byte[] frame, CancellationToken cancellationToken)
    {
        await _sendGate.WaitAsync(cancellationToken).ConfigureAwait(false);
        try
        {
            await _stream.WriteAsync(frame, cancellationToken).ConfigureAwait(false);
        }
        finally
        {
            _sendGate.Release();
        }
    }

    private static bool TryDecodePairingRequest(byte[] frame, out PairingRequest request)
    {
        try
        {
            var message = ProtocolJson.Decode(frame);
            if (message.Payload is PairingRequest value)
            {
                request = value;
                return true;
            }
        }
        catch (Exception error) when (error is JsonException or InvalidDataException)
        {
        }
        request = null!;
        return false;
    }

    public async ValueTask DisposeAsync()
    {
        if (Interlocked.Exchange(ref _disposed, 1) != 0) return;
        lock (_channelGate)
        {
            _secureChannel?.Dispose();
            _secureChannel = null;
        }
        _client.Dispose();
        _sendGate.Dispose();
        await ValueTask.CompletedTask;
    }
}
